designs/2021-suppression-support/README.md

```diff
@@ -0,0 +1,511 @@
+  }
+]
+```
+
+Figure Data Flow below is a high-level understanding of the current design.
+
+![ESLint](ESLint.png)
+
+- When rules are suppressed globally (external configuration files), ESLint does not do any tracking or run these suppressed rules at all, and suppression information will not be returned to output.
+- When rules are suppressed in-source (`eslint-disable` inside of a file), ESLint will still run the rules but then remove any lint messages that were triggered with the specified rule(s).
+
+### What we need
+
+As mentioned above, what we need is to allow engineers to provide justification in the suppression and output the suppression information.
+
+To minimize changes, we propose to use the `-- Description` part of the suppression as "Justification".
+
+More specifically, for inSource suppression (or inline directive comments in ESLint), the description after two or more consecutive dashes can be treated as the justification.
+
+For example:
+
+```javascript
+/* eslint-disable no-undef -- Justification example */
+a = 1;
+/* eslint-enable no-undef */
+```
+```javascript
+console.log("foo"); // eslint-disable-line no-console -- Justification example
+```
+```javascript
+// eslint-disable-next-line no-console -- Justification example
+console.log("bar");
+```
+
+Then we may get:
+
+```json
+{"kind": "inSource", "justification": "Justification example"}
+```
+
+If a violation is inline-suppressed multiple times, all suppression entries will be recorded, according to [suppressions property](https://docs.oasis-open.org/sarif/sarif/v2.0/csprd02/sarif-v2.0-csprd02.html#_Toc10127852). So `unusedDisableDirectives` might always be empty.
+
+For external suppression (or configuration files in ESLint), there are no good places to put a justification. We may put something general as a justification, like "Globally disabled".
+
+For example:
+
+```json
+{"kind": "external", "justification": "Globally disabled"}
+```
+
+If a line is suppressed by multiple rules, a suppression list would be preferred to gather all suppression information:
+
+```json
+suppressions: [
+  {"kind": "inSource", "justification": "Justification message 1"},
+  {"kind": "inSource", "justification": "Justification message 2"},
+  {"kind": "external", "justification": "Globally disabled"}
+]
+```
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+### Goal
+
+1.	Keep the current behavior of ESLint, i.e., errors/warnings will not be shown on the command line if the related rules are disabled, unless `--no-inline-config` is used.
+2.	When the new CLI option `--output-suppression` is used, the violations (problems) that disabled by inline comments will not be removed.
```

Comment:
Maybe `—track-suppressions` would be a better name because ESLint won’t output anything on its own?